File: src/main/kotlin/com/github/dxahtepb/etcdidea/view/BrowserToolWindow.java
package com.github.dxahtepb.etcdidea.view;

import com.github.dxahtepb.etcdidea.model.EtcdConnection;
import com.github.dxahtepb.etcdidea.model.EtcdKvEntries;
import com.github.dxahtepb.etcdidea.service.EtcdService;
import com.github.dxahtepb.etcdidea.view.actions.AddKeyAction;
import com.github.dxahtepb.etcdidea.view.actions.DeleteKeyAction;
import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.DefaultActionGroup;
import com.intellij.openapi.project.Project;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.ui.table.JBTable;

import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JPanel;
import javax.swing.JTextField;
import java.awt.BorderLayout;
import java.awt.event.ActionEvent;

public class BrowserToolWindow {
    private final EtcdService etcdService;
    private final Project project;

    private JPanel myToolWindowContent;

    private JTextField textField;
    private JButton getAllKeysButton;

    private JPanel tablePanel;
    private JPanel toolbarPanel;
    private EtcdTableModel resultsModel;
    private JBTable resultTable;

    public BrowserToolWindow(Project project, EtcdService etcdService) {
        this.etcdService = etcdService;
        this.project = project;
        initResultTable();
        initToolBar();
        initActionListeners();
    }

    public JPanel getContent() {
        return myToolWindowContent;
    }

    private void initActionListeners() {
        getAllKeysButton.addActionListener(this::buttonListener);
    }

    private void initToolBar() {
        toolbarPanel.setLayout(new BorderLayout());
        DefaultActionGroup actionResultGroup = new DefaultActionGroup("EtcdResultGroup", true);
        actionResultGroup.add(new AddKeyAction(this));
        actionResultGroup.add(new DeleteKeyAction(this));
        JComponent actionToolBarComponent = ActionManager.getInstance()
                .createActionToolbar("EtcdKvActionGroup", actionResultGroup, true)
                .getComponent();
        toolbarPanel.add(actionToolBarComponent, BorderLayout.WEST);
        toolbarPanel.setVisible(false);
    }

    private void initResultTable() {
        this.resultsModel = new EtcdTableModel();
        this.resultTable = new JBTable(resultsModel);
        resultTable.setDragEnabled(false);
        resultTable.getTableHeader().setReorderingAllowed(false);
        resultTable.setColumnSelectionAllowed(false);
    }

    private void buttonListener(ActionEvent event) {
        updateResults();
    }

    public void showAddKeyDialog() {
        new AddKeyDialogWindow(project, getCurrentConnection()).show();
        updateResults();
    }

    public void deleteSelectedKey() {
        int selectedRow = resultTable.getSelectedRow();
        if (selectedRow < 0) {
            return;
        }
        String selectedKey = (String) resultTable.getValueAt(selectedRow, 0);
        etcdService.deleteEntry(getCurrentConnection(), selectedKey);
        updateResults();
    }

    public boolean isRowSelected() {
        return resultTable.getSelectedRow() != -1;
    }

    private void updateResults() {
        EtcdKvEntries entries = etcdService.listAllEntries(getCurrentConnection());
        resultsModel.setDataVector(entries);
        displayResult(resultTable);
    }

    private void displayResult(JComponent tableView) {
        tablePanel.invalidate();
        tablePanel.removeAll();
        tablePanel.add(new JBScrollPane(tableView));
        tablePanel.validate();
        toolbarPanel.setVisible(true);
    }

    private EtcdConnection getCurrentConnection() {
        return new EtcdConnection(textField.getText());
    }
}

File: src/main/kotlin/com/github/dxahtepb/etcdidea/view/BrowserToolWindow.kt
package com.github.dxahtepb.etcdidea.view

import com.github.dxahtepb.etcdidea.model.EtcdServerConfiguration
import com.github.dxahtepb.etcdidea.service.EtcdService
import com.github.dxahtepb.etcdidea.view.actions.AddKeyAction
import com.github.dxahtepb.etcdidea.view.actions.AddServerAction
import com.github.dxahtepb.etcdidea.view.actions.DeleteKeyAction
import com.github.dxahtepb.etcdidea.view.actions.DeleteServerAction
import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.ActionToolbarPosition
import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.project.Project
import com.intellij.ui.AnActionButton
import com.intellij.ui.DoubleClickListener
import com.intellij.ui.OnePixelSplitter
import com.intellij.ui.ToolbarDecorator
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.table.JBTable
import com.intellij.ui.treeStructure.Tree
import com.intellij.util.ui.JBUI
import java.awt.BorderLayout
import java.awt.event.MouseEvent
import javax.swing.JComponent
import javax.swing.JPanel
import javax.swing.JTree
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.MutableTreeNode
import javax.swing.tree.TreePath
import javax.swing.tree.TreeSelectionModel.SINGLE_TREE_SELECTION

class BrowserToolWindow(private val project: Project, private val etcdService: EtcdService) {
    private lateinit var treeModel: DefaultTreeModel
    private lateinit var myTree: Tree

    private lateinit var resultsModel: EtcdTableModel
    private lateinit var resultTable: JBTable

    private val rootPanel: JPanel

    init {
        rootPanel = JPanel(BorderLayout()).apply {
            add(createToolbarPanel(), BorderLayout.NORTH)
            val splitterBorder = JBUI.Borders.customLine(JBUI.CurrentTheme.ToolWindow.borderColor(), 1, 0, 0, 0)
            add(
                OnePixelSplitter(true, 0.6f).apply {
                    firstComponent = createTreePanel()
                    secondComponent = createTablePanel().apply {
                        border = splitterBorder
                    }
                },
                BorderLayout.CENTER
            )
        }
    }

    fun getContent(): JComponent = rootPanel

    private fun createToolbarPanel(): JComponent {
        val actionGroup = DefaultActionGroup("BrowserActionGroup", false).apply {
            add(AddServerAction(::insertNewConfiguration))
            add(DeleteServerAction(::deleteSelectedConfiguration, ::isTreeSelected))
        }
        val actionToolbar = ActionManager.getInstance().createActionToolbar("EtcdBrowser", actionGroup, true)
        return JPanel(BorderLayout()).apply {
            add(actionToolbar.component, BorderLayout.CENTER)
        }
    }

    private fun createTreePanel(): JComponent {
        treeModel = DefaultTreeModel(DefaultMutableTreeNode())
        myTree = Tree(treeModel).apply {
            isEditable = true
            isRootVisible = false
            emptyText.text = "Add server configuration"
            selectionModel.selectionMode = SINGLE_TREE_SELECTION
        }

        object : DoubleClickListener() {
            override fun onDoubleClick(event: MouseEvent): Boolean {
                if (event.source !is JTree || !isTreeSelected()) return false
                updateResults()
                return true
            }
        }.installOn(myTree)

        return JPanel(BorderLayout()).apply {
            add(
                JBScrollPane(myTree).apply { border = JBUI.Borders.empty() },
                BorderLayout.CENTER
            )
        }
    }

    private fun insertNewConfiguration(configuration: EtcdServerConfiguration) {
        val childNode = DefaultMutableTreeNode(configuration, false)
        val root = treeModel.root as DefaultMutableTreeNode
        treeModel.insertNodeInto(childNode, root, root.childCount)
        myTree.scrollPathToVisible(TreePath(childNode.path))
    }

    private fun deleteSelectedConfiguration() {
        val nodeToRemove = myTree.selectionPath?.lastPathComponent as? MutableTreeNode ?: return
        treeModel.removeNodeFromParent(nodeToRemove)
    }

    private fun createTablePanel(): JComponent {
        resultsModel = EtcdTableModel()
        resultTable = JBTable(resultsModel).apply {
            dragEnabled = false
            tableHeader.reorderingAllowed = false
            columnSelectionAllowed = false
        }
        val resultTableDecorator = ToolbarDecorator.createDecorator(resultTable)
            .setToolbarPosition(ActionToolbarPosition.TOP)
            .setPanelBorder(JBUI.Borders.empty()).setScrollPaneBorder(JBUI.Borders.empty())
            .addExtraAction(AnActionButton.fromAction(AddKeyAction(this)))
            .addExtraAction(AnActionButton.fromAction(DeleteKeyAction(this)))
        return JPanel(BorderLayout()).apply {
            add(resultTableDecorator.createPanel(), BorderLayout.CENTER)
        }
    }

    fun showAddKeyDialog() {
        AddKeyDialogWindow(project, getCurrentConnection()).show()
        updateResults()
    }

    fun deleteSelectedKey() {
        val selectedRow = resultTable.selectedRow
        if (selectedRow < 0) return
        val selectedKey = resultTable.getValueAt(selectedRow, 0) as String
        etcdService.deleteEntry(getCurrentConnection(), selectedKey)
        updateResults()
    }

    private fun updateResults() {
        val entries = etcdService.listAllEntries(getCurrentConnection())
        resultsModel.setDataVector(entries)
    }

    fun isRowSelected() = resultTable.selectedRow != -1

    fun isTreeSelected() = myTree.selectionCount != 0

    private fun getCurrentConnection(): EtcdServerConfiguration {
        val node = myTree.selectionPath?.lastPathComponent as DefaultMutableTreeNode
        return node.userObject as EtcdServerConfiguration
    }
}
